### PR TITLE
feat: dfx new templates

### LINKED ISSUE #2771: `--no-frontend` flag
I tried the following:

1. `dfx new --no-frontend`

I expected to see this happen: No asset canister present, no entries in `dfx.json`

Instead, this happened: Asset canister and entrie present.

### Meta

`dfx --version`:
```
0.12.0
```

<details><summary>Further log output</summary>
<p>

```
<log output>
```

</p>
</details>

### PR DESCRIPTION
**Pull request by adamspofford-dfinity**
Also closes #2771.

### CODE REVIEW

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,14 @@
 
 # UNRELEASED
 
+### feat: new starter templates
+
+`dfx new` now has a new set of customizable project templates and an interactive menu for selecting them. Supports the Svelte, Vue, and React frameworks, and Azle and Kybra backends.
+
+### fix: --no-frontend no longer creates a frontend
+
+Previously --no-frontend still created a frontend canister. This behavior is now accessed via `--frontend simple-assets`.
```
Comment (sesi200) [inline CHANGELOG.md:11]:
```suggestion
Previously `dfx new --no-frontend` still created a frontend canister. This behavior is now accessed via `--frontend simple-assets`.
```
May be confusing without extra context

**Comment by krpeacock** [comment]
I think it's a problem that these templates won't run correctly with `dfx` up to the latest release version. I think we should add a `"build"` script to the root package.json that's compatible with dfx `<0.15.3`

File: src/dfx/assets/project_templates/react/src/__project_name___frontend/public/.ic-assets.json5 [hunk 2]
```diff
@@ -50,8 +50,7 @@
             // See: https://owasp.org/www-community/attacks/xss/
             "X-XSS-Protection": "1; mode=block"
         },
-        // Set the allow_raw_access field to false to redirect all requests from .raw.icp0.io to .icp0.io
-        // The default behavior is to allow raw access.
-        // "allow_raw_access": true
+        // redirect all requests from .raw.icp0.io to .icp0.io (this redirection is the default)
+        "allow_raw_access": false
```
[the remark this answers is not in the document]
Comment (adamspofford-dfinity) [reply src/dfx/assets/project_templates/react/src/__project_name___frontend/public/.ic-assets.json5:54]: The others came from @krpeacock's templates, and this one I was just making consistent with those. If we changed it away from this default for a reason I won't gainsay it.

**Comment by adamspofford-dfinity** [comment]
Is it? I don't picture a user generating a project with a newer version of dfx, then switching to an older one.

File: src/dfx/assets/project_templates/react/src/__project_name___frontend/public/.ic-assets.json5 (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericswanson-dfinity) [reply src/dfx/assets/project_templates/react/src/__project_name___frontend/public/.ic-assets.json5:54]: Yes.  The new templates should use the new default (and not specify the default in .ic_assets.json5)

**Comment by krpeacock** [comment]
> Is it? I don't picture a user generating a project with a newer version of dfx, then switching to an older one.

Someone might share their code, generated from a new version, and then someone could have trouble running it if the version isn't pinned or indicated. Probably an edge case though.

Still, having access to `"test", "start", and "build"` scripts is probably a nice thing to have, so that people don't have to `cd` or use the `--workspaces` flag for everything. (feedback from testing all the frontends manually)

**Comment by krpeacock** [comment]
image

Regression - the vanilla template has a different layout from the other versions

**Comment by krpeacock** [comment]
To fix, we can just delete the `App.css` from `vue` and `vanilla` and remove where it's imported. It's a holdover from Ryan's project. @adamspofford-dfinity do you want me to make the change?

**Comment by letmejustputthishere** [comment]
i know this is still WIP, but when attempting to create a new project, my frontend wasn't generated with the following output.

```
❯ sdk/target/debug/dfx new new_template
✔ Select a backend language: · Motoko
✔ Select a frontend framework: · Svelte
✔ Add extra features (space to select, enter to confirm) · Internet Identity, Frontend tests
Fetching manifest https://sdk.dfinity.org/manifest.json
  Version v0.16.0+rev30.f7b611f1 installed successfully.
Creating new project "new_template"...
CREATE       new_template/dfx.json (179B)...
CREATE       new_template/.gitignore (260B)...
CREATE       new_template/README.md (2.47KiB)...
CREATE       new_template/tsconfig.json (280B)...
CREATE       new_template/package.json (477B)...
CREATE       new_template/src/new_template_backend/main.mo (105B)...
WARN: Node could not be found. Skipping installing the frontend example code.
WARN: You can bypass this check by using the --frontend flag.
CREATE       new_template/src/new_template_frontend/assets/sample-asset.txt (24B)...
CREATE       new_template/src/new_template_frontend/assets/.ic-assets.json5 (5.37KiB)...
Creating git repository...
```

i can confirm that `node` indeed wasn't available in this shell, but if `node` is needed to install `dfx` we should add a prerequisites section [here](https://internetcomputer.org/docs/current/developer-docs/setup/install/).

**Comment by letmejustputthishere** [comment]
PS: i think this is amazing, it would be even cooler if templates with II would [integrate with the frontend code](https://github.com/letmejustputthishere/vite-sveltekit-motoko/blob/main/src/routes/%2Bpage.svelte) as well.

File: src/dfx/src/commands/new.rs [hunk 3]
```diff
@@ -63,6 +66,77 @@ pub struct NewOpts {
...
+            Self::Azle => "TypeScript (Azle)",
+            Self::Kybra => "Python (Kybra)",
+        }
+        .fmt(f)
+    }
+}
+
+#[derive(ValueEnum, Debug, Copy, Clone, PartialEq, Eq)]
+enum FrontendType {
+    Svelte,
```
Comment (letmejustputthishere) [inline src/dfx/src/commands/new.rs:96]: to be more accurate, can we use sveltekit instead of svelte here?

**Comment by letmejustputthishere** [comment]
support for e2e tests using `pocket-ic` would be amazing as well 🤩

**Review by krpeacock — APPROVED** [review]
From my perspective, this is ready to go out (once all the checks pass)

**Comment by krpeacock** [comment]
it would be even cooler if templates with II would [integrate with the frontend code](https://github.com/letmejustputthishere/vite-sveltekit-motoko/blob/main/src/routes/%2Bpage.svelte) as well.

I do plan to do this, but it'll be a follow-up

File: src/dfx/src/commands/new.rs (hunk 3 above)
Comment (ericswanson-dfinity) [reply src/dfx/src/commands/new.rs:95]: @krpeacock do you have an opinion on this?

File: src/dfx/assets/project_templates/react/src/__project_name___frontend/public/.ic-assets.json5 (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericswanson-dfinity) [reply src/dfx/assets/project_templates/react/src/__project_name___frontend/public/.ic-assets.json5:54]: These changes to defaults are the only reason I still have changes requested on this PR

File: src/dfx/src/commands/new.rs (hunk 3 above)
Comment (ericswanson-dfinity) [reply src/dfx/src/commands/new.rs:95]: @adamspofford-dfinity In the absence of an opinion from anyone with a better handle on the ecosystem, I think we should change this to SvelteKit (and the cli option to `--sveltekit`), since there is a distinction between the two.

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

**Review by ericswanson-dfinity — APPROVED** [review]

File: e2e/tests-dfx/new.bash [hunk 4]
```diff
@@ -51,9 +52,34 @@ teardown() {
   assert_eq "motoko"
 }
 
-@test "dfx new always emits sample-asset.txt" {
-  assert_command dfx new e2e_frontend --frontend
-  assert_file_exists e2e_frontend/src/e2e_frontend_frontend/assets/sample-asset.txt
-  assert_command dfx new e2e_no_frontend --no-frontend
-  assert_file_exists e2e_no_frontend/src/e2e_no_frontend_frontend/assets/sample-asset.txt
+@test "frontend templates apply successfully" {
+  for frontend in sveltekit vue react vanilla simple-assets none; do
```
Comment (ericswanson-dfinity) [inline e2e/tests-dfx/new.bash:56]: This may be what is breakinf
```suggestion
  for frontend in svelte-kit vue react vanilla simple-assets none; do
```
